Added basic git rename-branch

ricmatsui opened the pull request:
@kevgo @charlierudolph @allewun 
Still need to support undo correctly. Closes #474 

Is there a way to generate the man pages? Or are they created manually?

ricmatsui commented on features/git-rename-branch/non_feature_branch/without_remote_origin/without_open_changes.feature:
```diff
@@ -0,0 +1,36 @@
+  Scenario: error when trying to rename
+    When I run `git rename-branch production renamed-production`
+    Then it runs no Git commands
+    And I get the error "The branch 'production' is not a feature branch. Run 'git rename-branch production renamed-production -f' to force the rename, then reconfigure git-town on any other clones of this repo."
```
Is there a good way to break this up other than `cat <<EOF` or `"\ <CR>"`?

kevgo replied:
What exactly do you want to break up here? The error message string? I think we should do that. It is important to see both error messages (1. the branch is not a feature branch, 2. guidance on how to rename it anyways), but the spec shouldn't care in which order they happen.

```cucumber
And I get the error "The branch 'production' is not a feature branch."
And I get the error "Run 'git rename-branch production renamed-production -f' to force the rename, then reconfigure git-town on any other clones of this repo."

kevgo commented on documentation/commands/git-rename-branch.md:
```diff
@@ -0,0 +1,37 @@
+* pushes the new branch
+* deletes the old branch from the remote repository
+Deletes the old branch locally
+Reconfigures git-town locally if renaming a non-feature branch
```
Lets follow this format for this paragraph:
* longer general description of the command
* edge cases and other things worth knowing
* list of the steps executed

This means

```
Renames the given branch on both the local machine and the remote if one is configured. 
This command is intended for feature branches. Renaming non-feature branches has to be confirmed with the `-f` option.

[ list of steps here ]
```

kevgo commented on documentation/commands/git-rename-branch.md:
```diff
@@ -0,0 +1,37 @@
+    The name of the branch to rename.
+
+<newbranchname>
+    The new name of the branch being renamed.
```
The new name for the branch

?

kevgo commented on features/git-rename-branch/destination_branch_exists_locally.feature:
```diff
@@ -0,0 +1,30 @@
+
+  As a developer trying to rename a branch with the name of an existing branch
+  I should see an error telling me that a branch with that name already exists
+  So that my new feature branch is unique.
```
As a developer trying to rename a branch into an already existing branch
I want to the command to abort with an error message
So that I don't lose work by accidentally overwriting existing branches.

kevgo commented on features/git-rename-branch/destination_branch_exists_locally.feature:
```diff
@@ -0,0 +1,30 @@
+
+
+  Background:
+    Given I have a feature branches named "current-feature" and "existing-feature"
```
Given I have ~~a~~ **the** feature branches...

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,34 @@
+
+  As a developer with a poorly named feature branch
+  I want to be able to rename it safely in one easy step
+  So that I can stay organized and remain productive
```
So that I can keep my branches well organized and remain productive

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,34 @@
+    Given I have a feature branch named "current-feature"
+    And the following commits exist in my repository
+      | BRANCH          | LOCATION         | MESSAGE        |
+      | main            | local and remote | main commit    |
```
Do we really need the main commit here?

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,34 @@
+      | BRANCH          | COMMAND                                         |
+      | current-feature | git fetch --prune                               |
+      |                 | git stash -u                                    |
+      |                 | git checkout -b renamed-feature current-feature |
```
You could also perform the local rename through `git branch -m <oldname> <newname>`.

ricmatsui replied:
Yeah, the reason I didn't use `-m` is because I think it would make it more complicated and asymmetric since the remote branches don't have a `-m` equivalent. It also lets this reuse existing code, and doesn't have a significant difference in performance.

kevgo replied:
Okay, sounds good to me then!

kevgo commented on features/git-rename-branch/destination_branch_exists_locally.feature:
```diff
@@ -0,0 +1,30 @@
+    And I get the error "A branch named 'existing-feature' already exists"
+    And I am still on the "current-feature" branch
+    And I still have my uncommitted file
+
```
And I am left with my original commits

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/without_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,30 @@
+    And I have the following commits
+      | BRANCH          | LOCATION         | MESSAGE        |
+      | main            | local and remote | main commit    |
+      | renamed-feature | local            | feature commit |
```
Cool feature that it doesn't push the branch up here if it wasn't pushed before!

kevgo commented on features/git-rename-branch/non_feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,44 @@
+  Scenario: error when trying to rename
+    When I run `git rename-branch production renamed-production`
+    Then it runs no Git commands
+    And I get the error "The branch 'production' is not a feature branch. Run 'git rename-branch production renamed-production -f' to force the rename, then reconfigure git-town on any other clones of this repo."
```
The name for this tool currently is `Git Town`, not `git-town`. I'm happy either way, we just should be consistent.

kevgo commented on features/git-rename-branch/non_feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,44 @@
+      |                    | git branch -D production                      |
+      |                    | git stash pop                                 |
+    And I end up on the "renamed-production" branch
+    And my non-feature branches are now configured as "qa" and "renamed-production"
```
That's cool!

kevgo commented on features/git-rename-branch/not_enough_branch_names.feature:
```diff
@@ -0,0 +1,27 @@
+    Then it runs no Git commands
+    And I get the error "No branch name provided"
+    And I am still on the "current-feature" branch
+
```
And I am left with my original commits

kevgo commented on features/git-rename-branch/on_main_branch.feature:
```diff
@@ -0,0 +1,23 @@
+    When I run `git rename-branch main renamed-main -f`
+    Then it runs no Git commands
+    And I get the error "The main branch cannot be renamed."
+    And I am still on the "main" branch
```
I think we should also specify the scenario when somebody wants to rename a branch that doesn't exist.

kevgo replied:
And make sure that if one renames a branch onto itself, it still works correctly.

kevgo replied:
Also, we should specify the behavior for renaming a branch that contains unpushed or unpulled changes. None of these should get lost when renaming a branch, otherwise this command isn't safe.

ricmatsui replied:
Good points.

ricmatsui replied:
Resolved.

kevgo commented on src/git-rename-branch:
```diff
@@ -0,0 +1,71 @@
+    ensure_has_branch "$target_branch_name"
+  fi
+
+  echo $force_rename
```
Remove this debug statement

kevgo commented on src/git-rename-branch:
```diff
@@ -0,0 +1,71 @@
+    ensure_is_feature_branch "$target_branch_name" "Run 'git rename-branch $target_branch_name $destination_branch_name -f' to force the rename, then reconfigure git-town on any other clones of this repo."
+  fi
+
+  if [ -z "$destination_branch_name" ]; then
```
This could be added to line 21, where you check the other args.

ricmatsui replied:
I think the errors should be different to be clear that two branch names are required, but I can move it up closer to the other.

kevgo commented on src/git-rename-branch:
```diff
@@ -0,0 +1,71 @@
+  if [ "$(is_non_feature_branch "$target_branch_name")" = true ]; then
+    local new_branches=$(remove_string "$NON_FEATURE_BRANCH_NAMES" ',' "$target_branch_name")
+    new_branches=$(insert_string "$new_branches" ',' "$destination_branch_name")
+    store_configuration non-feature-branch-names "$new_branches"
```
I think lines 62-64 can be extracted into a helper method.

kevgo commented:
Seriously cool feature, @ricmatsui! Branch names encounter a good amount of technical drift, and being able to rename them at any time will be super useful! Thanks for taking this on!

kevgo commented:
The man pages are created manually at this point. We were thinking about generating them somehow, but haven't gotten around to invest the time into finding tool support and setting all this up. Writing man pages by hand is interesting though, not something one does every day, and a great chance to see how they work.

ricmatsui commented:
Sounds good, thanks. I have addressed the comments and I think we can have it abort if the tracking branch is not in sync for now so no work is lost. In a later pull request we can add merging/rebasing automatically as well.

Do you think implementing undo by `rename-branch`ing again would work? One of the steps is a push, so it won't undo completely using the `undo_steps_for`.

kevgo commented on documentation/commands/git-rename-branch.md:
```diff
@@ -0,0 +1,45 @@
+
+Renames the given branch on both the local machine and the remote if one is configured.
+Aborts if the new branch name already exists or the tracking branch is out of sync.
+This command is intended for feature branches. Renaming non-feature branches has to be confirmed with the `-f` option.
```
This is really well written!

ricmatsui replied:
You did recommend it that way :wink:

kevgo replied:
Really? Sorry, I didn't mean to celebrate my own ideas here, hahaha   :smile:

kevgo commented on features/git-rename-branch/branch_does_not_exist.feature:
```diff
@@ -0,0 +1,28 @@
+
+  As a developer mistyping the feature branch name to rename
+  I should get an error that the given branch does not exist
+  So that I can rename the correct branch
```
So that I can try again with the correct branch name.

kevgo commented on features/git-rename-branch/branch_does_not_exist.feature:
```diff
@@ -0,0 +1,28 @@
+  Background:
+    Given the following commits exist in my repository
+      | BRANCH | LOCATION         | MESSAGE     |
+      | main   | local and remote | main commit |
```
I don't think we need this commit for this feature spec.

kevgo commented on features/git-rename-branch/destination_branch_exists_locally.feature:
```diff
@@ -0,0 +1,36 @@
+
+  As a developer trying to rename a branch into an already existing branch
+  I want the command to abort with an error message
+  So that I don't lose work by accidentally overwriting existing branches
```
Please add a dot at the end, since this is an (admittedly pretty lengthy and cumbersome) sentence.

kevgo commented on features/git-rename-branch/destination_branch_exists_remotely.feature:
```diff
@@ -0,0 +1,29 @@
+    And I get the error "A branch named 'existing-feature' already exists"
+    And I am still on the "current-feature" branch
+    And I still have my uncommitted file
+
```
And I am left with my original commits

?

kevgo commented on features/git-rename-branch/destination_branch_exists_remotely.feature:
```diff
@@ -0,0 +1,29 @@
+      | BRANCH          | COMMAND           |
+      | current-feature | git fetch --prune |
+    And I get the error "A branch named 'existing-feature' already exists"
+    And I am still on the "current-feature" branch
```
same here, `And I am left with my original commits`?

ricmatsui replied:
Resolved

kevgo commented on features/git-rename-branch/feature_branch/branch_and_destination_same.feature:
```diff
@@ -0,0 +1,29 @@
+
+  As a developer renaming a feature branch onto itself
+  I should get a message saying no action is needed
+  So that it works as expected 
```
So that I am aware that I just did a no-op.

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_unpulled_changes.feature:
```diff
@@ -0,0 +1,33 @@
+Feature: git rename-branch: errors if renaming a feature branch that has unpulled changes
+
+  As a developer renaming a feature branch that has unpulled changes
```
Reading this again, it becomes clearer to me that we definitely must implement the branch syncing into this command (in a separate PR). Otherwise, without running `git fetch`, how do we even know whether this branch has unpulled changes?

ricmatsui replied:
Agreed. It does run `git fetch`, but it currently does not do a merge.

kevgo commented on features/git-rename-branch/non_feature_branch/branch_and_destination_same.feature:
```diff
@@ -0,0 +1,30 @@
+
+  As a developer renaming a non-feature branch onto itself
+  I should get a message saying no action is needed
+  So that it works as expected 
```
So that I know I just did a no-op.

kevgo commented on features/git-rename-branch/on_main_branch.feature:
```diff
@@ -0,0 +1,23 @@
+Feature: git rename-branch: errors with renaming the main branch
```
replace "with~ with "when"

kevgo commented:
I think making this undoable would be very usefule and consistent. And it could easily be made possible if we implement renaming branches following these steps:
* rename the branch locally
* remove existing tracking branch (if one exists)
* add (the new) tracking branch
* add the new non-feature branch name
* remove the old non-feature branch name

Each of these commands has a simple and easy way of undoing it, which can de defined in `undo_steps_for`.

What do you think?

ricmatsui commented:
Yes, sorry I wasn't very clear. The issue I'm having is `push` has this undo step:
```bash
function undo_steps_for_push {
  echo "skip_current_branch_steps $UNDO_STEPS_FILE"
}
```
Which basically causes the destination branch to still exist locally and remotely after the undo if there is a tracking branch since when the rename occurs the destination branch is checked out and pushed. What do you think of having another `reversible_push` command or something similar that doesn't have this undo step?

kevgo commented:
Yes, I think we need to separate the push command into two: one called `create_tracking_branch` that creates a new remote branch (and is undone by deleting the remote branch), and one that simply pushes updates to an already existing tracking branch. Which of these command is used has to be determined based on whether the tracking branch already exists. 

The fact that both commands are implemented differently in Git is another indicator that these really are two different commands. 

I have created #519 to track and implement this separately from this change. I think we can implement the undo for git-rename-branch in a separate PR as well, to keep things simple and manageable here.

kevgo commented on features/git-rename-branch/branch_does_not_exist.feature:
```diff
@@ -0,0 +1,28 @@
+
+  As a developer mistyping the feature branch name to rename
+  I should get an error that the given branch does not exist
+  So that I can rename the correct branch name.
```
I think one doesn't rename names, one renames branches here. 
`So that I can rename the correct branch.`

kevgo commented on features/git-rename-branch/destination_branch_exists_locally.feature:
```diff
@@ -0,0 +1,36 @@
+Feature: git rename-branch: errors when the destination branch exists locally
+
+  As a developer trying to rename a branch into an already existing branch
```
Does one rename into, or rename to? I think rename to.

`As a developer trying to rename a branch to an already existing branch`

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,34 @@
+
+  As a developer with a poorly named feature branch
+  I want to be able to rename it safely in one easy step
+  So that I can keep my branches well organized and remain productive.
```
What do you think about

`So that the names of my branches match what they implement, can I can manage them effectively.`

kevgo commented on features/git-rename-branch/feature_branch/with_remote_origin/with_tracking_branch/with_unpushed_changes.feature:
```diff
@@ -0,0 +1,33 @@
+
+  As a developer renaming a feature branch that has unpushed changes
+  I should get an error that the given branch is not in sync with its tracking branch
+  So that I don't lose work by accidentally overwriting existing branches.
```
What do you think about 

`So that I don't lose work by deleting branches that contain commits that haven't been pulled yet.`

ricmatsui replied:
Sure, I'll put this in `with_unpulled_changes.feature`

kevgo replied:
Whoops, yes, good catch. Here we can then say what you said in the other comment ("that I know branches must be in sync in order to be renamed").

kevgo commented on features/git-rename-branch/non_feature_branch/with_remote_origin/with_tracking_branch/with_open_changes.feature:
```diff
@@ -0,0 +1,45 @@
+
+  As a developer with a poorly named non-feature branch
+  I want to be able to rename it safely in one easy step
+  So that I can keep my branches well organized and remain productive.
```
Please use the same version here as in the spec for renaming feature branches.

kevgo commented on features/git-rename-branch/non_feature_branch/with_remote_origin/with_tracking_branch/with_unpulled_changes.feature:
```diff
@@ -0,0 +1,34 @@
+
+  As a developer renaming a non-feature branch that has unpulled changes
+  I should get an error that the given branch is not in sync with its tracking branch
+  So that I don't lose work by accidentally overwriting existing branches.
```
Let's use the same line here as in the corresponding feature spec for renaming feature branches.

kevgo commented on features/git-rename-branch/non_feature_branch/with_remote_origin/with_tracking_branch/with_unpushed_changes.feature:
```diff
@@ -0,0 +1,34 @@
+
+  As a developer renaming a non-feature branch that has unpushed changes
+  I should get an error that the given branch is not in sync with its tracking branch
+  So that I don't lose work by accidentally overwriting existing branches.
```
Technically we wouldn't lose changes if the branch to rename contains unpushed changes. I'd be open to just go ahead with the rename in this case.

ricmatsui replied:
Yeah, I suppose it would just sync the branch. We could implement some detection for only unpushed changes, but I think it would only be needed until syncing is added, so I don't know if it's worth it. What do you think of a correction to `So that I know the branch must be in sync` or similar?

kevgo replied:
You are right. Sounds good!  :)

kevgo commented:
Okay, this looks good to go from my end once the remaining few comments are addressed and if you decide to implement the proposed change in behavior in a separate PR. Nice work!

kevgo commented:
:rocket:
